src/state/slices/txHistorySlice/txHistorySlice.ts

```diff
@@ -142,23 +174,77 @@ const updateOrInsert = (txHistory: TxHistory, tx: Tx, accountSpecifier: AccountS
   // get applied to state when it goes out of scope
 }
 
+type UpdateOrInsertRebase = (txState: TxHistory, data: RebaseHistoryPayload['payload']) => void
+
+const updateOrInsertRebase: UpdateOrInsertRebase = (txState, payload) => {
```

Comment:
Just a thought, but we could write this logic in a more pure way (note, not tested):

```ts
const updateOrInsertRebase: UpdateOrInsertRebase = (txState, payload) => {
  const { accountId, assetId } = payload
  let { rebases } = txState
  rebases = (() => {
    const byId: RebaseById = payload.data.map(rebase => {
      const rebaseId = makeRebaseId({ accountId, assetId, rebase })
      return { [rebaseId]: rebase }
    })

    const ids = payload.data.reduce((acc, rebase) => {
      const rebaseId = makeRebaseId({ accountId, assetId, rebase })
      const isNew = !txState.rebases.byId[rebaseId]
      if (isNew) {
        const orderedRebases = orderBy(rebases.byId, 'blockTime', ['desc'])
        const index = orderedRebases.findIndex(
          rebase => makeRebaseId({ accountId, assetId, rebase }) === rebaseId
        )
        return { ...acc, ids: slice(rebases.ids, index).concat(rebaseId) }
      } else {
        return acc
      }
    })

    type Rebase = {
      byAssetId: RebaseByAssetId
      byAccountId: RebaseByAccountId
      ids: RebaseId[]
      byId: RebaseById
    }

    const byAssetId: RebaseByAssetId = payload.data.reduce((acc: Rebase, rebase: Rebase) => {
      acc.byAssetId[assetId] = addToIndex(
        rebases.ids,
        rebases.byAssetId[assetId],
        makeRebaseId({ accountId, assetId, rebase })
      )
      return acc
    }).byAssetId

    const byAccountId: RebaseByAccountId = payload.data.reduce((acc: Rebase, rebase: Rebase) => {
      acc.byAccountId[assetId] = addToIndex(
        rebases.ids,
        rebases.byAccountId[accountId],
        makeRebaseId({ accountId, assetId, rebase })
      )
      return acc
    }).byAccountId

    return {
      ...rebases,
      byId,
      ids,
      byAssetId,
      byAccountId
    }
  })()
}
```

Reply:
i really like this pattern, and am going to steal it along with try/catch IIFE style, but if i update this i want to update the txHistory upsert too, and don't want to fix regressions here. i know the slice logic changes looks suspect at first glance (i know you said untested here)

Reply:
It's also slightly less performant, so not always the best choice - but generally, I prefer it as it makes spotting bugs and understanding what's going on way easier.

Yeah, I'm not sure the reducer quite works and didn't wanna waste time on it, but it's something like that.